utils/src/task_manager/mod.rs

```diff
@@ -0,0 +1,406 @@
+    }
+
+    /// Check if the cluster shutdown condition is met
+    fn check(&self) -> bool {
+        let mpmc_channel_shutdown = self.mpmc_channel_shutdown.load(Ordering::Relaxed);
+        let sync_follower_task_count = self.sync_follower_task_count.load(Ordering::Relaxed);
+        let leader_notified = self.leader_notified.load(Ordering::Relaxed);
+        mpmc_channel_shutdown && sync_follower_task_count == 0 && leader_notified
+    }
+}
+
+impl TaskManager {
+    /// Create a new `TaskManager`
+    #[must_use]
+    #[inline]
+    pub fn new() -> Self {
+        let tasks = Arc::new(DashMap::new());
+        for name in TaskName::iter() {
+            let task = Task::new(name);
+            _ = tasks.insert(name, task);
+        }
+        for (from, to) in ALL_EDGES {
+            _ = tasks.get_mut(&from).map(|mut t| t.depend_by.push(to));
+            _ = tasks
+                .get_mut(&to)
+                .map(|mut t| t.depend_cnt = t.depend_cnt.overflow_add(1));
+        }
+        let state = Arc::new(AtomicU8::new(0));
+        let cluster_shutdown_tracker = Arc::new(ClusterShutdownTracker::new());
+        Self {
+            tasks,
+            state,
+            cluster_shutdown_tracker,
+        }
+    }
+
+    /// Check if task manager is shutdown
+    #[must_use]
+    #[inline]
+    pub fn is_shutdown(&self) -> bool {
+        self.state.load(Ordering::Acquire) != 0
+    }
+
+    /// Get shutdown listener
+    #[must_use]
+    #[inline]
+    pub fn get_shutdown_listener(&self, name: TaskName) -> Listener {
+        let task = self
+            .tasks
+            .get(&name)
+            .unwrap_or_else(|| unreachable!("task {:?} should exist", name));
+        Listener::new(
+            Arc::clone(&self.state),
+            Arc::clone(&task.notifier),
+            Arc::clone(&self.cluster_shutdown_tracker),
+        )
+    }
+
+    /// Spawn a task
+    #[inline]
+    pub fn spawn<FN, F>(&self, name: TaskName, f: FN)
+    where
+        F: Future<Output = ()> + Send + 'static,
+        FN: FnOnce(Listener) -> F,
+    {
+        if self.is_shutdown() {
+            return;
+        }
+        info!("spawn {name:?}");
+        let mut task = self
+            .tasks
+            .get_mut(&name)
+            .unwrap_or_else(|| unreachable!("task {:?} should exist", name));
+        let listener = Listener::new(
+            Arc::clone(&self.state),
+            Arc::clone(&task.notifier),
+            Arc::clone(&self.cluster_shutdown_tracker),
+        );
+        let handle = tokio::spawn(f(listener));
+        task.handle.push(handle);
+    }
+
+    /// Get root tasks queue
+    fn root_tasks_queue(tasks: &DashMap<TaskName, Task>) -> VecDeque<TaskName> {
+        tasks
+            .iter()
+            .filter_map(|task| (task.depend_cnt == 0).then_some(task.name))
+            .collect()
+    }
+
+    /// Inner shutdown task
+    async fn inner_shutdown(tasks: Arc<DashMap<TaskName, Task>>, state: Arc<AtomicU8>) {
```

Comment:
Please add an unit test to cover this method, make sure that the topology order is always as you expected.